File: packages/browser/catalog.mjs
import { Plugin, PluginGroup, makePlugin, makeRule } from "./plugin.mjs";
import { data } from "./identity.mjs";
export function loadCatalog(runtime, catalog) {
  const groups = [];
  for (const definition of catalog.groups) {
    let group =
      definition.parent === null
        ? new PluginGroup()
        : groups[definition.parent].group();
    if (definition.name !== null) group = group.named(definition.name);
    groups.push(group);
  }
  const rules = catalog.rules.map((definition, index) =>
    makeRule(runtime, index, definition),
  );
  const plugins = catalog.plugins.map((definition) => {
    const plugin =
      definition.group === null ? new Plugin() : groups[definition.group].new();
    const state = data(plugin, "plugin");
    return makePlugin({
      ...state,
      name: definition.name,
      frozen: true,
      rules: definition.rules.map((index) => rules[index]),
    });
  });
  function tree(value, leaf) {
    if (typeof value === "number") return leaf(value);
    return Object.freeze(
      Object.fromEntries(
        Object.entries(value).map(([name, child]) => [name, tree(child, leaf)]),
      ),
    );
  }
  const presets = new Map();
  return {
    plugins: tree(catalog.exports.plugins, (index) => plugins[index]),
    presets: tree(catalog.exports.presets, (index) => {
      if (!presets.has(index)) {
        const definition = catalog.presets[index];
        presets.set(
          index,
          runtime.preset(
            {
              plugins: definition.plugins.map((index) =>
                data(plugins[index], "plugin"),
              ),
              order: definition.order,
            },
            definition,
          ),
        );
      }
      return presets.get(index);
    }),
  };
}

File: crates/markdown/src/bindings/catalog.rs
use super::{Composition, GroupSpec, PluginSpec};
use crate::{
    GrammarBuilder,
    ast::ExtensionDefinition,
    engine::{Contribution, Plugin, PluginGroup},
};
use serde::Serialize;
use serde_json::Value;

#[derive(Serialize)]
pub struct RuleSpec {
    pub name: Option<String>,
    pub phase: &'static str,
    pub extensions: Vec<&'static str>,
}
#[derive(Serialize)]
pub struct PresetSpec {
    pub description: String,
    pub extensions: Vec<&'static str>,
    pub plugins: Vec<usize>,
    pub order: Vec<usize>,
}
#[derive(Default)]
pub struct Catalog {
    pub(super) groups: Vec<PluginGroup>,
    pub(super) plugins: Vec<Plugin>,
    pub(super) rules: Vec<Contribution>,
    pub(super) presets: Vec<PresetSpec>,
    pub exports: Value,
}
impl Catalog {
    fn group(&mut self, group: &PluginGroup) -> usize {
        if let Some(index) = self.groups.iter().position(|g| g.same(group)) {
            return index;
        }
        if let Some(parent) = group.parent() {
            self.group(parent);
        }
        self.groups.push(group.clone());
        self.groups.len() - 1
    }
    pub(crate) fn plugin(&mut self, plugin: &Plugin) -> usize {
        if let Some(index) = self.plugins.iter().position(|p| p.same(plugin)) {
            return index;
        }
        if let Some(group) = plugin.namespace() {
            self.group(group);
        }
        for rule in plugin.rules.iter() {
            if !self.rules.iter().any(|r| r.same(rule)) {
                self.rules.push(rule.clone());
            }
        }
        self.plugins.push(plugin.clone());
        self.plugins.len() - 1
    }
    pub(crate) fn preset(&mut self, definition: &GrammarBuilder) -> usize {
        // Metadata registration validates the recipe without compiling dispatch
        // tables or invoking a preset parser for recognizer warmup.
        let extensions = definition
            .validate()
            .expect("valid bundled preset")
            .iter()
            .map(|extension| extension.name)
            .collect();
        let description = definition.describe();
        let plugins = definition.plugins.iter().map(|p| self.plugin(p)).collect();
        let order = definition
            .rules
            .iter()
            .map(|(_, r)| self.rule_index(r))
            .collect();
        self.presets.push(PresetSpec {
            plugins,
            order,
            description,
            extensions,
        });
        self.presets.len() - 1
    }
    fn rule_index(&self, rule: &Contribution) -> usize {
        self.rules
            .iter()
            .position(|r| r.same(rule))
            .expect("registered rule")
    }
    pub fn preset_composition(&self, index: usize) -> Option<Composition> {
        let preset = self.presets.get(index)?;
        Some(Composition {
            groups: self.group_specs(),
            plugins: preset
                .plugins
                .iter()
                .map(|&i| self.plugin_spec(&self.plugins[i]))
                .collect(),
            order: preset.order.clone(),
        })
    }
    pub fn extensions(&self) -> impl Iterator<Item = &ExtensionDefinition> {
        self.rules.iter().flat_map(Contribution::extensions)
    }
    fn group_specs(&self) -> Vec<GroupSpec> {
        self.groups
            .iter()
            .map(|g| GroupSpec {
                parent: g.parent().map(|p| {
                    self.groups
                        .iter()
                        .position(|g| g.same(p))
                        .expect("registered parent")
                }),
                name: g.name().map(str::to_owned),
            })
            .collect()
    }
    fn plugin_spec(&self, plugin: &Plugin) -> PluginSpec {
        PluginSpec {
            group: plugin.namespace().map(|g| {
                self.groups
                    .iter()
                    .position(|p| p.same(g))
                    .expect("registered namespace")
            }),
            name: plugin.name().map(str::to_owned),
            rules: plugin.rules.iter().map(|r| self.rule_index(r)).collect(),
        }
    }
    pub fn describe(&self) -> Value {
        let rules = self
            .rules
            .iter()
            .map(|r| RuleSpec {
                name: r.name().map(str::to_owned),
                phase: r.phase(),
                extensions: r.extensions().iter().map(|d| d.name).collect(),
            })
            .collect::<Vec<_>>();
        let plugins = self
            .plugins
            .iter()
            .map(|p| self.plugin_spec(p))
            .collect::<Vec<_>>();
        serde_json::json!({
            "groups": self.group_specs(), "plugins": plugins, "rules": rules,
            "presets": self.presets, "exports": self.exports,
        })
    }
}
